server/src/main/java/org/elasticsearch/monitor/fs/FsHealthService.java

```diff
@@ -0,0 +1,189 @@
+    }
+
+     class FsHealthMonitor implements Runnable {
+
+        static final String TEMP_FILE_NAME = ".es_temp_file";
+        private byte[] byteToWrite;
+
+        FsHealthMonitor(){
+            this.byteToWrite = new byte[20];
+        }
+
+        @Override
+        public void run() {
+            try {
+                if (enabled) {
+                    monitorFSHealth();
+                    logger.debug("Monitor for FS health ran successfully {}", pathHealthStats);
+                }
+            } catch (Exception e) {
+                logger.error("Exception monitoring FS health", e);
+            }
+        }
+
+        private void monitorFSHealth() {
+            for (Path path : nodeEnv.nodeDataPaths()) {
+                long executionStartTime = currentTimeMillisSupplier.getAsLong();
+                slowPathHealthCheckTimeoutHandler = threadPool.schedule(new Runnable() {
+                    @Override
+                    public void run() {
+                        logSlowDataPath(Level.WARN, path, executionStartTime);
+                    }
+                    @Override
+                    public String toString() {
+                        return "Scheduled timeout for logging slow IO " + FsHealthService.FsHealthMonitor.this;
+                    }
+                }, slowPathLoggingThreshold, ThreadPool.Names.GENERIC);
+
+                try {
+                    if (Files.exists(path)) {
+                        Path tempDataPath = path.resolve(TEMP_FILE_NAME);
+                        Files.deleteIfExists(tempDataPath);
+                        try (OutputStream os = Files.newOutputStream(tempDataPath, StandardOpenOption.CREATE_NEW)) {
+                            new Random().nextBytes(byteToWrite);
+                            os.write(byteToWrite);
+                            IOUtils.fsync(tempDataPath, false);
+                        }
+                        Files.delete(tempDataPath);
+                        pathHealthStats.put(path, Status.HEALTHY);
+                    }
+                } catch (Exception ex) {
+                    logger.error("Failed to perform FS health check on path {} due to {} ", path, ex);
+                    pathHealthStats.put(path,  Status.UNHEALTHY);
+
+                }
+                cancelTimeoutHandler();
+            }
+            lastRunTimeMillis.getAndUpdate(l -> currentTimeMillisSupplier.getAsLong());
+        }
+
+         private void cancelTimeoutHandler() {
+             if (slowPathHealthCheckTimeoutHandler != null) {
+                 slowPathHealthCheckTimeoutHandler.cancel();
+             }
+         }
+
+         private void logSlowDataPath(Level level, Path path, long executionStartTime) {
+             final TimeValue elapsedTime = TimeValue.timeValueMillis(currentTimeMillisSupplier.getAsLong() - executionStartTime);
+             logger.log(level, "Slow IO detected on path [{}], FS health check elapsed time [{}]", path, elapsedTime);
```

Comment:
Suggest a message of the form `checking writability of [{}] took [{}ms] which is above the warn threshold of [{}]` since this is the wording used in other similar messages.